subworkflow cis:
    workdir: "../data/cis"
    snakefile: "../data/cis/Snakefile"

subworkflow imfusion:
    workdir: "../data/rnaseq_imfusion"
    snakefile: "../data/rnaseq_imfusion/Snakefile"

genes = ['Erg', 'Ets1', 'Pten', 'Xrcc6', 'Nanos3', 'Rapgef6', 'Rps3']

rule all:
    input:
        "plot_aneup.pdf",
        expand("{gene}.pdf", gene=genes)

rule hit_table:
    input:
        rscript = "cis_sanger.r",
        dset = cis("cis_per_tumor.RData")
    output:
        table = "cis_sanger.tsv"
    shell:
        "Rscript {input.rscript} --infile {input.dset} --outfile {output.table}"

rule cis:
    input:
        table = "cis_sanger.tsv",
        reference = imfusion("STAR_GRCm38_ens92_PB")
    output:
        assocs = "cis_sanger_results.tsv"
    shell:
        "imfusion-ctg"
            " --insertions {input.table}"
            " --reference {input.reference}"
            " --output {output.assocs}"

rule plot_aneup:
    input:
        rscript = "plot_aneup.r",
        insertions = "cis_sanger.tsv",
        dna = "cis_sanger_results.tsv",
        rna = imfusion("merged_ctgs.txt"),
        exons = imfusion("exon_counts.txt")
    output:
        plot = "plot_aneup.pdf"
    shell:
        "Rscript {input.rscript}"
            " --insertions {input.insertions}"
            " --dna {input.dna}"
            " --rna {input.rna}"
            " --exons {input.exons}"
            " --plotfile {output.plot}"

rule plot_insert:
    input:
        rscript = "plot_fusions.r",
        ctgs = imfusion("merged_ctgs.txt"),
        insertions = imfusion("insertions.txt")
    output:
        plot = "{gene}.pdf"
    shell:
        "Rscript {input.rscript}"
            " --ctgs {input.ctgs}"
            " --insertions {input.insertions}"
            " --gene {wildcards.gene}"
            " --plotfile {output.plot}"
